### ISSUE TITLE
http: proxy error: http: invalid Read on closed Body

### ISSUE THREAD

**Issue by soamvasani** [fission/fission#707]
From slack user thenamly:

Fission workflows call the function and fail with router returning 502. This is sometimes happening but not always:
```
Invoking Fission function: 'http://router.fission/fission-function/datainput'." component=fnenv.fission fn="runtime:\"fission\" ID:\"datainput\" "
[datainput][status]: 502" component=fnenv.fission fn="runtime:\"fission\" ID:\"datainput\" "
[datainput][Content-Type]:  " component=fnenv.fission fn="runtime:\"fission\" ID:\"datainput\" "
[datainput][output]: ''" component=fnenv.fission fn="runtime:\"fission\" ID:\"datainput\" "
msg="[%!s(int=502)] Failed : %!v(MISSING)" component=fnenv.fission fn="runtime:\"fission\" ID:\"datainput\" "
```
Then rest of the invocation completely fail:
```
"Backend client appending event." aggregate.id=DataInput aggregate.type=function event.id= event.type=TASK_FAILED nats.Subject=invocation.wi-bb481bba-9470-46a1-b5d8-8be973e688db
"Handling invocation notification!" component=controller-wi labels="map[aggregate.type:function parent.type:invocation parent.id:wi-bb481bba-9470-46a1-b5d8-8be973e688db event.id:5 event.type:TASK_FAILED aggregate.id:DataInput]" notification=TASK_FAILED
msg="Scheduler evaluating..." component=scheduler wfi=wi-bb481bba-9470-46a1-b5d8-8be973e688db
msg="Applying action: fail" component=controller-wi
msg="Backend client appending event." aggregate.id=wi-bb481bba-9470-46a1-b5d8-8be973e688db aggregate.type=invocation event.id= event.type=INVOCATION_FAILED nats.Subject=invocation.wi-bb481bba-9470-46a1-b5d8-8be973e688db
msg="Handling invocation notification!" component=controller-wi labels="map[aggregate.id:wi-bb481bba-9470-46a1-b5d8-8be973e688db aggregate.type:invocation parent.type: parent.id: event.id:6 event.type:INVOCATION_FAILED]" notification=INVOCATION_FAILED
I  10.16.0.16 - - [28/May/2018:11:40:30 +0000] "POST / HTTP/1.1" 500 7
msg="Invocation not successful, was 'FAILED'"
 ```
fission router gives these two after that:
```
request to datainput-bio4wmgi.fission-function errored out. backing off for 50ms before retrying
http: proxy error: http: invalid Read on closed Body
 
```

**Comment by thenamly** [comment]
@soamvasani This issue still happening even on the latest version of the fission.

We are facing this issue a lot when we have higher concurrency to the same function 'a' with executor type new deploy. 

Here is the more detailed summary of logs:

1. executor: Error: 404: key 'f426065e-9a7f-11e8-b854-42010a9c014f_3172885' not found // this is same uid + resource version of the pod of 'a'
1. router: Calling getServiceForFunction for function: a 
1. router: Calling getServiceForFunction for function: a
1. router http: proxy error: Resource not found - key 'f426065e-9a7f-11e8-b854-42010a9c014f_3172885' not found // same key of 'a' as above
1. a: // logs that shows that one pod of 'a' is working now
1. executor: [a] Checking for cached function service
1. executor: [a] No cached function service found, creating one
1. executor: [a] getting environment for function
1. executor: k8s object printed <Name:a UID:f426065e-9a7f-11e8-b854-42010a9c014f ResourceVersion:3172885 // the same a pod
1. executor: [a] Checking for cached function service
1. executor: Waiting for concurrent request for the same function: // the same k8s object with the same uid and resource version printed here
1. executor: "POST /v2/getServiceForFunction HTTP/1.1" 404 61
1. executor: "POST /v2/getServiceForFunction HTTP/1.1" 200 41
1. fission function call returns 502 at the end

It feels like there is a concurrency bug on executor. This is the major bug we have and we sometimes get 60% error rate due to this bug.

What do you think? where this bug coming from?

**Comment by thenamly** [comment]
@erwinvaneyk We solved horizontal scalability problem of workflows and now we hit this error a lot when we increase the concurrency.

It doesn't happen to only single function 'a'. It happens to different functions all the time.

**Comment by soamvasani** [comment]
Okay, the 404 from executor is really strange.  Since you're reproing this regularly, can you send us logs from executor, router and controller, around the time this occurs?

**Comment by soamvasani** [comment]
How many router instances are you running, btw?

**Comment by soamvasani** [comment]
> executor: Waiting for concurrent request for the same function: // the same k8s object with the same uid and resource version printed here
> executor: "POST /v2/getServiceForFunction HTTP/1.1" 404 61

I seem to recall @life1347 talking about a known bug related to that wait in the first line, but I can't remember whether it's already fixed... Ta-Ching?

**Comment by life1347** [comment]
@thenamly  I'm a little bit confused by the summary of log

From the summary, looks like the logs combined the logs from `poolmgr` and `newdeploy`. It normally occurs when the executortype was `poolmgr` originally and changed to `newdeploy`. 

The log 1, 4, 12 are for the same request. But what made me confused is the summary

```
a: // logs that shows that one pod of 'a' is working now
```
Is the raw log similar to following log?

```
[map[executorInstanceId:mmqarsgg unmanaged:true functionName:json-upload functionUid:11654455-5cda-11e8-a3dd-6c92bf6f7932]] found 3 ready pods of 3 total
```

If yes, then at this moment the executortype of function was `poolmgr`. The detailed reason can be found [here](https://github.com/fission/fission/issues/695#issuecomment-398822419)

If not, could you paste raw log for debugging? Thanks!

And the rest logs, they are for function with executortype `newdeploy`. 

In my previous tests, you will get 502 when a request is sent to the pod that is not ready to serve requests due to the process of specialization is not finished yet.

@thenamly Could you help me to do a test with following limitation:
1. Executor type: newdeploy
2. minScale = 3 & maxScale = 3

Then send requests after 1 or 2 mins of function creation, then check whether you still get tons of 502 errors. If rate of 502 erros is lower than before, then that means the impletation of newdeploy need to improve to prevent pod receives requests before specialization is finished.

**Comment by thenamly** [comment]
I am not visiting github regularly, sorry for rate reply (you can ping me on slack if there is a need for quciker reply).

@soamvasani what I send is actually filtered/annotated version of all logs coming from cluster but I can send full logs in a few days.

@soamvasani I run only single router and executor type is newdeploy for all functions

@life1347 for your deleted reply. Our min scale is 1 max is 100. I can try 3-3 if you still want.

@life1347 Happy to try your router improvements under load. Just glancing the diff of PR, as a side comment, please don't put any low timeout to function execution. Our functions (doing AI stuff) have 30 to 100ish seconds work that needs to be done sometimes. 

To summarize things I can help here:
* Send full logs again (I can skip this one if not required)
* Try 3-3 (I think I don't need to try that anymore)
* Try fission bundle on that branch (I am happy to try this with quick concurrency test and see the error rate. If I am getting less errors, try again with higher concurrency)

PS: @life1347 Completely irrelevant but I'll have 1 month long business trip soon to Taiwan and happy to meet you for a coffee over there if you wish. I'll ping you on slack

**Comment by thenamly** [comment]
Changelog of v0.10 is missing but probably #752 will be helpful for debugging (as in the #695). Probably, makes sense to try to repro problem on new version

**Comment by life1347** [comment]
@thenamly  Yes, you're right. 

Currently, when router encounters HTTP request problems it will remove function service from the cache.
```
request to languagerecognizer-4hxjtbjy.fission-function errored out. removing function : languagerecognizer from router's cache and requesting a new service for function
```
And once the record was removed, router will try to get new function service record from executor
```
Calling getServiceForFunction for function: languagerecognizer
```
Every new request to router will make it to send a request to executor, that mean if router is under high workload there might tons of requests try to get new record from executor and caused 502 error.
So in PR https://github.com/fission/fission/pull/856 , I add a update lock to router so that there at most one goroutine can send "get new function service" request to exectuor to prevent 502 error.

**Comment by thenamly** [comment]
This is what happened when I use your optimized router although there are less 502 errors, they still 5-10% of the requests.

@life1347 you can see new logs below

```
request to abc-phen5jcg.fission-function errored out. backing off for 100ms before retrying
 Add update lock for function speechrecognizer
Calling getServiceForFunction for function: speechrecognizer 
Remove update lock for function speechrecognizer
http: proxy error: Error sending request to function speechrecognizer: http: invalid Read on closed Body
```

**Comment by life1347** [comment]
Solved by #963
